proposals/TypeSelection.md

```diff
@@ -0,0 +1,149 @@
+# Type selection
+
+
+The idea of this proposal is to describe a way to be able to do `.` selection over types. 
+
+*For example*
+
+```dataweave
+type User = {name: String}
+
+type NameType = User.name
+```
+
+With this option it will allow us to reference a type from any object with out the need to refactoring.
+
+## Supported selections
+
+There are two kind of selections static and dynamic selections. 
+
+### Static selection
+
+For static selections we should support both with and without namesapce. A question would be if we want to do the same matching. 
+
+*For example*
+
+```dataweave
+ns ns1 http://acme.com
+
+type User = {
+  ns1#name : String,
+  name : Number,
+}
+
+type A = User.name
+```
+
+If the type selection follows the runtime semantics type `A` should be String, but I think that it should return `Number`. This may introduce some noice but I think is much clear.
```

Comment:
Another option is to return the union of all possible outcomes, in this case `String | Number`. I think this is worth evaluating, since our runtime is not guided by the type of a variable, but by its contents, which don't need to be on the same order:

```%dw 2.0
output application/json
ns x https://google.com
var a: {x#b: String, b: Number} = {b: 1, x#b: "dss"}
var b: {x#b: String, b: Number} = {x#b: "dss", b: 1}

---
b.b // String
// a.b Number
```

Reply:
If we want to follow same semantics as value selector then yes it should return a Union but my question is should we? In the past we took the decision of not forcing namespace because it forces users to add namespaces on every selection something that is done a lot in a script (and to be honest I'm not sure was the right call). But in this case where we are talking about type expression (something that should be less frequent). I'm not sure that same logic applies. So in your example 
```
output application/json
ns x https://google.com

type TypeA = {x#b: String, b: Number} 
type TypeB =  {x#b: String, b: Number} 

var a: TypeA = {b: 1, x#b: "dss"}
var b: TypeB= {x#b: "dss", b: 1}

var x: TypeA.b = b.b //This shouldn't check as this selection may return String | Number and TypeA.b returns Number

var z: TypeA.b | TypeA.x#b= b.b //This will check

var z1: TypeA.x#b= b.x#b //This will check
---
b.b // String
// a.b Number
```